File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/httpdir/CachingCatalog.java
package app.zxtune.fs.httpdir;

import android.content.Context;

import androidx.annotation.VisibleForTesting;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import app.zxtune.TimeStamp;
import app.zxtune.fs.dbhelpers.CommandExecutor;
import app.zxtune.fs.dbhelpers.FileTree;
import app.zxtune.fs.dbhelpers.QueryCommand;
import app.zxtune.fs.dbhelpers.Timestamps;

final class CachingCatalog implements Catalog {

  private final static TimeStamp DIR_TTL = TimeStamp.fromDays(1);

  private final RemoteCatalog remote;
  private final FileTree db;
  private final CommandExecutor executor;

  CachingCatalog(Context ctx, RemoteCatalog remote, String id) {
    this(remote, new FileTree(ctx, id), id);
  }

  @VisibleForTesting
  CachingCatalog(RemoteCatalog remote, FileTree db, String id) {
    this.remote = remote;
    this.db = db;
    this.executor = new CommandExecutor(id);
  }

  @Override
  public void parseDir(final Path path, final DirVisitor visitor) throws IOException {
    final String dirName = path.getLocalId();
    executor.executeQuery("dir", new QueryCommand() {

      private final Timestamps.Lifetime lifetime = db.getDirLifetime(dirName, DIR_TTL);

      @Override
      public boolean isCacheExpired() {
        return lifetime.isExpired();
      }

      @Override
      public void updateCache() throws IOException {
        final ArrayList<FileTree.Entry> entries = new ArrayList<>(100);
        remote.parseDir(path, new DirVisitor() {
          @Override
          public void acceptDir(String name, String desc) {
            entries.add(new FileTree.Entry(name, desc, null));
          }

          @Override
          public void acceptFile(String name, String descr, String size) {
            entries.add(new FileTree.Entry(name, descr, size));
          }
        });
        db.runInTransaction(() -> {
          db.add(dirName, entries);
          lifetime.update();
        });
      }

      @Override
      public boolean queryFromCache() {
        final List<FileTree.Entry> entries = db.find(dirName);
        if (entries != null) {
          for (FileTree.Entry e : entries) {
            if (e.isDir()) {
              visitor.acceptDir(e.name, e.descr);
            } else {
              visitor.acceptFile(e.name, e.descr, e.size);
            }
          }
          return true;
        } else {
          return false;
        }
      }
    });
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/httpdir/CachingCatalog.kt
package app.zxtune.fs.httpdir

import android.content.Context
import androidx.annotation.VisibleForTesting
import app.zxtune.TimeStamp
import app.zxtune.fs.dbhelpers.CommandExecutor
import app.zxtune.fs.dbhelpers.FileTree
import app.zxtune.fs.dbhelpers.QueryCommand
import app.zxtune.fs.httpdir.Catalog.DirVisitor
import java.util.*

private val DIR_TTL = TimeStamp.fromDays(1)

class CachingCatalog @VisibleForTesting constructor(
    private val remote: RemoteCatalog,
    private val db: FileTree,
    id: String
) : Catalog {
    private val executor: CommandExecutor = CommandExecutor(id)

    constructor(ctx: Context, remote: RemoteCatalog, id: String) : this(
        remote,
        FileTree(ctx, id),
        id
    )

    override fun parseDir(path: Path, visitor: DirVisitor) {
        val dirName = path.getLocalId()
        executor.executeQuery("dir", object : QueryCommand {
            private val lifetime = db.getDirLifetime(dirName, DIR_TTL)
            override val isCacheExpired: Boolean
                get() = lifetime.isExpired

            override fun updateCache() {
                val entries = ArrayList<FileTree.Entry>(100)
                remote.parseDir(path, object : DirVisitor {
                    override fun acceptDir(name: String, description: String) {
                        entries.add(FileTree.Entry(name, description, null))
                    }

                    override fun acceptFile(name: String, description: String, size: String) {
                        entries.add(FileTree.Entry(name, description, size))
                    }
                })
                db.runInTransaction {
                    db.add(dirName, entries)
                    lifetime.update()
                }
            }

            override fun queryFromCache() = db.find(dirName)?.run {
                forEach {
                    // TODO: use extracted FileTree.Entry.feed
                    if (it.isDir) {
                        visitor.acceptDir(it.name, it.descr)
                    } else {
                        visitor.acceptFile(it.name, it.descr, it.size)
                    }
                }
                true
            } ?: false
        })
    }
}
